File: frontend/src/api/gridTrading.ts
import request from './request'

export interface GridSignal {
  symbol: string
  name: string
  action: 'BUY' | 'SELL' | 'HOLD' | 'NO_QUOTE'
  quantity: number
  current_price: number | null
  anchor_price: number | null
  anchor_source?: string | null
  last_grid_price: number | null
  trigger_price: number | null
  next_buy_price?: number
  next_sell_price?: number
  grid_pct: number | null
  grid_level?: number
  position_qty?: number
  available_qty?: number
  base_position_qty?: number
  reason: string
  timestamp: string
  signal_key?: string
  order_preview?: Record<string, unknown> | null
}

export interface GridStatus {
  xtdata_available: boolean
  xttrader_available: boolean
  quote_connected: boolean
  account_configured: boolean
  account_id: string
  account_type: string
  trader_path: string
  data_dir?: string | null
  order_submit_enabled: boolean
  error?: string | null
}

export interface GridAccount {
  cash: number
  total_asset: number
  market_value: number
  positions: Record<string, Record<string, unknown>>
  source: string
  error?: string | null
}

export interface GridSignalsResponse {
  timestamp: string
  symbols: string[]
  account: {
    cash: number
    total_asset: number
    market_value: number
    source: string
    error?: string | null
  }
  quote_error?: string | null
  order_submit_enabled: boolean
  signals: GridSignal[]
}

export interface GridSignalRequest {
  params?: Record<string, unknown>
  manual_account?: Record<string, unknown> | null
}

export interface TechSmallCapVariant {
  key: string
  name: string
  description: string
  params: Record<string, unknown>
}

export interface TechSmallCapOrder {
  symbol: string
  side: 'BUY' | 'SELL'
  quantity: number
  reference_price?: number
  price_type?: string
  strategy_name?: string
  remark?: string
}

export interface TechSmallCapSignalsResponse {
  timestamp: string
  strategy: string
  variant: string
  trade_date: string
  account: {
    cash: number
    total_asset: number
    market_value: number
    source: string
    error?: string | null
  }
  universe_size: number
  candidate_count: number
  target_symbols: string[]
  target_weights: Record<string, number>
  entry_filter: Record<string, unknown>
  quote_error?: string | null
  order_submit_enabled: boolean
  orders: TechSmallCapOrder[]
  top_candidates: Record<string, unknown>[]
  excluded_symbol_count: number
}

export const gridTradingApi = {
  status: () =>
    request.get<GridStatus>('/grid-trading/status'),
  account: () =>
    request.get<GridAccount>('/grid-trading/account'),
  signals: (data: GridSignalRequest = {}) =>
    request.post<GridSignalsResponse>('/grid-trading/signals', data),
  submitPreview: (orderPreview: Record<string, unknown>) =>
    request.post<Record<string, unknown>>('/grid-trading/orders/preview', { order_preview: orderPreview }),
  submitOrder: (order: Record<string, unknown>) =>
    request.post<Record<string, unknown>>('/grid-trading/orders/submit', { order }),
  techSmallCapVariants: () =>
    request.get<TechSmallCapVariant[]>('/grid-trading/tech-small-cap/variants'),
  techSmallCapSignals: (data: GridSignalRequest = {}) =>
    request.post<TechSmallCapSignalsResponse>('/grid-trading/tech-small-cap/signals', data),
  submitTechSmallCapOrders: (orders: Record<string, unknown>[], confirm = false) =>
    request.post<Record<string, unknown>>('/grid-trading/tech-small-cap/orders/submit', { orders, confirm }),
}

File: backend/app/api/grid_trading.py
"""Grid trading signal API."""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.grid_trading import GRID_SYMBOLS, grid_trading_service
from app.services.tech_small_cap_live import tech_small_cap_live_service

router = APIRouter(prefix="/grid-trading", tags=["grid-trading"])


class GridSignalRequest(BaseModel):
    params: dict[str, Any] = Field(default_factory=dict)
    manual_account: dict[str, Any] | None = None


class GridSubmitPreviewRequest(BaseModel):
    order_preview: dict[str, Any] = Field(default_factory=dict)


class GridSubmitOrderRequest(BaseModel):
    order: dict[str, Any] = Field(default_factory=dict)


class TechSmallCapSignalRequest(BaseModel):
    params: dict[str, Any] = Field(default_factory=dict)
    manual_account: dict[str, Any] | None = None


class TechSmallCapSubmitOrdersRequest(BaseModel):
    orders: list[dict[str, Any]] = Field(default_factory=list)
    confirm: bool = False


@router.get("/symbols")
async def get_grid_symbols():
    return {"code": 0, "data": {"symbols": GRID_SYMBOLS}}


@router.get("/status")
async def get_grid_status():
    data = await grid_trading_service.status()
    return {"code": 0, "data": data}


@router.get("/account")
async def get_grid_account():
    account = await grid_trading_service.account()
    return {
        "code": 0,
        "data": {
            "cash": account.cash,
            "total_asset": account.total_asset,
            "market_value": account.market_value,
            "positions": account.positions,
            "source": account.source,
            "error": account.error,
        },
    }


@router.post("/signals")
async def get_grid_signals(req: GridSignalRequest):
    data = await grid_trading_service.signals(
        params=req.params,
        manual_account=req.manual_account,
    )
    return {"code": 0, "data": data}


@router.post("/orders/preview")
async def submit_grid_order_preview(req: GridSubmitPreviewRequest):
    data = await grid_trading_service.submit_order_preview(req.order_preview)
    return {"code": 0, "data": data}


@router.post("/orders/submit")
async def submit_grid_order(req: GridSubmitOrderRequest):
    data = await grid_trading_service.submit_order(req.order)
    return {"code": 0, "data": data}


@router.get("/tech-small-cap/variants")
async def get_tech_small_cap_variants():
    return {"code": 0, "data": tech_small_cap_live_service.variants()}


@router.post("/tech-small-cap/signals")
async def get_tech_small_cap_signals(req: TechSmallCapSignalRequest):
    data = await tech_small_cap_live_service.signals(
        params=req.params,
        manual_account=req.manual_account,
    )
    return {"code": 0, "data": data}


@router.post("/tech-small-cap/orders/submit")
async def submit_tech_small_cap_orders(req: TechSmallCapSubmitOrdersRequest):
    data = await tech_small_cap_live_service.submit_orders(req.orders, confirm=req.confirm)
    return {"code": 0, "data": data}
